DSN. Add pieces-by-sector publishing and receiving.

shamil-gadelshin opened the pull request:
This PR adds a publication for the pieces from archival storage (sectors or L1). It also adds a placeholder to a handler for extracting pieces from sectors. Fixes: https://github.com/subspace/subspace/issues/882  fixes https://github.com/subspace/subspace/issues/867

### Changes
- Add piece retrieval from cold storage.
- farmer: Add piece-by-sector publishing.
- farmer: Refactor piece receiver.
- networking: Add a multihash type for sector and piece index hash.
- networking: Refactor multimess module.
- farmer: Add cancellation support for piece retrieval.

### Comments
- we need to implement an actual handler for the piece retrieval from the sector
- piece removal from publication workflow should be made along with sector restructuring

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](../CONTRIBUTING.md)

shamil-gadelshin committed:
networking: Refactor multimess module.

- rename to multihash module and fix usages

shamil-gadelshin committed:
Add archival storage pieces publishing.

- refactor benches
- refactor piece receiving on the farmer side
- add sector-piece publishing
- add sector-piece receiving

nazar-pc commented on crates/subspace-farmer/src/single_disk_plot.rs:
```diff
@@ -698,6 +701,22 @@ impl SingleDiskPlot {
+                            // Publish pieces-by-sector if we use DSN
+                            if let Some(ref dsn_node) = dsn_node {
+                                let piece_publisher =
+                                    PieceSectorPublisher::new(dsn_node.clone(), &shutting_down);
```
Can we create it outside of the loop once?

i1i1 replied:
Should it be constructed once and used each time?

nazar-pc commented on crates/subspace-farmer/src/single_disk_plot/piece_publisher.rs:
```diff
@@ -0,0 +1,74 @@
+        pieces_indexes: Vec<PieceIndex>,
+    ) -> Result<(), Box<dyn Error + Send + Sync + 'static>> {
+        for piece_index in pieces_indexes {
+            'attempts: loop {
```
Just FYI, https://crates.io/crates/backoff is ideally a better option, but for MVP this will do as well

nazar-pc commented on crates/subspace-farmer/src/single_disk_plot/piece_receiver.rs:
```diff
@@ -0,0 +1,236 @@
+
+pub struct BenchPieceReceiver {
+    piece: Piece,
+}
```
Let's not have impls before corresponding definition and please move this into benches, we ideally shouldn't expose bench- and test-specific stuff publicly if possible and in this case it is certainly possible.

nazar-pc commented on crates/subspace-farmer/src/single_disk_plot/piece_receiver.rs:
```diff
@@ -0,0 +1,236 @@
+                    return Ok(Some(piece));
+                }
+
+                sleep(Duration::from_secs(GET_PIECE_WAITING_DURATION_IN_SECS)).await;
```
Let's at least print a warning so application doesn't appear to be stuck completely.

nazar-pc reviewed (commented):
Minor comments, otherwise makes sense for this stage

shamil-gadelshin committed:
farmer: Fix review comments.

- refactored benchmarks
- added more logging
- optimized usage of the PiecePublisher

shamil-gadelshin committed:
Merge branch 'main' into dsn-sector-piece-provider

shamil-gadelshin committed:
Merge branch 'main' into dsn-sector-piece-provider
